### PR TITLE
add primitive head

### PR DESCRIPTION
**Pull request by encore-zhou**

### CODE REVIEW

File: mmdet3d/models/dense_heads/primitive_head.py [hunk 1]
```diff
@@ -0,0 +1,816 @@
...
+            newd = torch.sum(para_points * plane_lower_temp[:3], 1)
+            if self.check_upright(para_points) and \
+                    plane_lower_temp[0] + plane_lower_temp[1] < \
+                    self.train_cfg['lower_thresh']:
+                plane_lower = torch.as_tensor([0, 0, 1, plane_lower_temp[-1]
+                                               ]).to(device)
+                plane_upper = torch.as_tensor([0, 0, 1,
+                                               -torch.mean(newd)]).to(device)
+            else:
+                import pdb
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/primitive_head.py:357]: debug codes should be cleaned

File: mmdet3d/models/dense_heads/primitive_head.py [hunk 2]
```diff
@@ -0,0 +1,816 @@
...
+        return (para_points[0][-1] == para_points[1][-1]) and (
+            para_points[1][-1]
+            == para_points[2][-1]) and (para_points[2][-1]
+                                        == para_points[3][-1])
+
+    def check_z(self, plane_equ, para_points):
+        return torch.sum(para_points[:, 2] +
+                         plane_equ[-1]) / 4.0 < self.train_cfg['lower_thresh']
+
+    def get_linesel(self, points, xmin, xmax, ymin, ymax):
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/primitive_head.py:709]: `get_linesel` and `get_linesel2` are not good names. `get_lower_linesel` and `get_upper_linesel` might be better.

File: mmdet3d/models/dense_heads/primitive_head.py [hunk 3]
```diff
@@ -0,0 +1,816 @@
...
+            line_sel1, line_sel2, line_sel3, line_sel4 = self.get_linesel(
+                x[sel], xmin, xmax, ymin, ymax)
+            if self.primitive_mode == 'line':
+                if torch.sum(line_sel1) > self.train_cfg['num_point_line']:
+                    point_line_mask[ind[sel][line_sel1]] = 1.0
+                    linecenter = torch.mean(x[sel][line_sel1], axis=0)
+                    linecenter[1] = (ymin + ymax) / 2.0
+                    point_line_offset[
+                        ind[sel][line_sel1]] = linecenter - x[sel][line_sel1]
+                    point_line_sem[ind[sel][line_sel1]] = torch.as_tensor([
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/primitive_head.py:380]: For all the usage of ` torch.as_tensor(data).to(device)`, we can use `points.new_tensor(data)` for simipilicity

File: mmdet3d/models/dense_heads/primitive_head.py [hunk 4]
```diff
@@ -0,0 +1,816 @@
...
+
+        Returns:
+            tuple[torch.Tensor]: Targets of primitive head.
+        """
+        gt_bboxes_3d = gt_bboxes_3d.to(points.device)
+        num_points = points.shape[0]
+        device = points.device
+
+        if self.primitive_mode == 'z':
+            point_boundary_mask_z = torch.zeros(num_points).to(device)
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/primitive_head.py:315]: For all the usage of ` torch.zeros(data).to(device)`, we can use `points.new_zeros(data)` for simipilicity

File: mmdet3d/models/dense_heads/primitive_head.py [hunk 5]
```diff
@@ -0,0 +1,816 @@
...
+                    sel_global = ind[sel]
+                    point_boundary_mask_z[sel_global] = 1.0
+                    point_boundary_sem_z[sel_global] = torch.as_tensor([
+                        center[0], center[1], center[2], xmax - xmin,
+                        ymax - ymin, (pts_semantic_mask[ind][0])
+                    ]).to(device)
+                    point_boundary_offset_z[sel_global] = center - x[sel]
+
+            # Get left two lines
+            v1 = corners[3] - corners[2]
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/primitive_head.py:499]: v1, v2, cp, d, a, b, c are not good variable names. Use more clear names with regarding to their real usages and meaning.

File: mmdet3d/models/dense_heads/primitive_head.py [hunk 6]
```diff
@@ -0,0 +1,816 @@
...
+                    point_boundary_offset_xy)
+        elif self.primitive_mode == 'line':
+            return (point_line_mask, point_line_sem, point_line_offset)
+        else:
+            NotImplementedError
+
+    def primitive_decode_scores(self, net, end_points, num_class, mode=''):
+        net_transposed = net.transpose(2, 1)  # (batch_size, 1024, ..)
+
+        base_xyz = end_points['aggregated_points_' +
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/primitive_head.py:684]: change
```python

base_xyz = end_points['aggregated_points_' +
                              mode]  # (batch_size, num_proposal, 3)
```
to 
```python
# (batch_size, num_proposal, 3)
base_xyz = end_points['aggregated_points_' + mode]
```

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 7]
```diff
@@ -0,0 +1,821 @@
...
+from mmdet3d.models.builder import build_loss
+from mmdet3d.models.model_utils import VoteModule
+from mmdet3d.ops import PointSAModule, furthest_point_sample
+from mmdet.core import multi_apply
+from mmdet.models import HEADS
+
+
+@HEADS.register_module()
+class PrimitiveHead(nn.Module):
+    r"""Bbox head of `H3dnet <https://arxiv.org/abs/2006.05682>`_.
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:14]: H3DNet

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 8]
```diff
@@ -0,0 +1,821 @@
...
+            newd = torch.sum(para_points * plane_left_temp[:3], 1)
+            if plane_left_temp[2] < self.train_cfg['lower_thresh']:
+                plane_left = plane_left_temp
+                plane_right = points.new_tensor([
+                    plane_left_temp[0], plane_left_temp[1], plane_left_temp[2],
+                    -torch.mean(newd)
+                ])
+            else:
+                raise NotImplementedError
+                # print('error with upright')
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:540]: Remove this line

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 9]
```diff
@@ -0,0 +1,821 @@
...
+            newd = torch.sum(para_points * plane_front_temp[:3], 1)
+            if plane_front_temp[2] < self.train_cfg['lower_thresh']:
+                plane_front = plane_front_temp
+                plane_back = points.new_tensor([
+                    plane_front_temp[0], plane_front_temp[1],
+                    plane_front_temp[2], -torch.mean(newd)
+                ])
+            else:
+                raise NotImplementedError
+                # print('error with upright')
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:654]: Remove this line as well as unnecessary comments.

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 10]
```diff
@@ -0,0 +1,821 @@
...
+            Tuple: Loss of primitive module.
+        """
+        batch_size = primitive_center.shape[0]
+        vote_xyz_reshape = primitive_center.view(batch_size * num_proposal, -1,
+                                                 3)
+        center_loss = self.center_loss(
+            vote_xyz_reshape,
+            gt_primitive_center,
+            dst_weight=gt_primitive_mask.view(batch_size * num_proposal, 1))[1]
+        center_loss = center_loss.sum() / (
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:800]: The `loss.sum()/xxx` is unnecessary, we can set `reduction='sum'` and pass `torch.sum(gt_primitive_mask.float()) + 1e-6` as avg_factor to the loss to simplify the code, so as other losses.

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 11]
```diff
@@ -0,0 +1,821 @@
...
+                dst_weight=gt_primitive_mask.view(batch_size * num_proposal,
+                                                  1))[1]
+            size_loss = size_loss.sum() / (
+                torch.sum(gt_primitive_mask.float()) + 1e-6)
+        else:
+            size_loss = torch.tensor(0).float().to(center_loss.device)
+
+        # Semantic cls loss
+        sem_cls_loss = self.semantic_loss(semantic_scores, gt_sem_cls_label)
+        sem_cls_loss = torch.sum(sem_cls_loss * gt_primitive_mask.float()) / (
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:818]: These reduction can be achieved in a simpler way.

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 12]
```diff
@@ -0,0 +1,781 @@
...
+        """Generate primitive center from primitive head predictions.
+
+        Args:
+            pred_flag (torch.Tensor): Scores of primitive center.
+            center (torch.Tensor): Pridictions of primitive center.
+
+        Returns:
+            Tuple: Primitive center and the prediction indices.
+        """
+        ind_normal = self.softmax_normal(pred_flag)
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:774]: `self.softmax_normal(pred_flag)` -> `F.softmax(pred_flag, dim=1)`

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 13]
```diff
@@ -0,0 +1,781 @@
...
+                    bias=True,
+                    inplace=True))
+            prev_channel = feat_channels[k]
+        self.conv_pred = nn.Sequential(*conv_pred_list)
+
+        conv_out_channel = 3 + num_dim + num_classes
+        self.conv_pred.add_module('conv_out',
+                                  nn.Conv1d(prev_channel, conv_out_channel, 1))
+
+        self.softmax_normal = nn.Softmax(dim=1)
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:112]: This is unnecessary

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 14]
```diff
@@ -0,0 +1,781 @@
...
+            # use fps in vote_aggregation
+            sample_indices = None
+        elif sample_mod == 'seed':
+            # FPS on seed and choose the votes corresponding to the seeds
+            sample_indices = furthest_point_sample(seed_points,
+                                                   self.num_proposal)
+        elif sample_mod == 'random':
+            # Random sampling from the votes
+            batch_size, num_seed = seed_points.shape[:2]
+            sample_indices = seed_points.new_tensor(
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:157]:
```python
sample_indices = torch.randint(0, num_seed, (batch_size, self.num_proposal), device=seed_points.device)

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 15]
```diff
@@ -0,0 +1,781 @@
...
+            para_points = cur_corners[[1, 2, 5, 6]]
+            newd = torch.sum(para_points * plane_lower_temp[:3], 1)
+            if self.check_upright(para_points) and \
+                    plane_lower_temp[0] + plane_lower_temp[1] < \
+                    self.train_cfg['lower_thresh']:
+                plane_lower = points.new_tensor(
+                    [0, 0, 1, plane_lower_temp[-1]])
+                plane_upper = points.new_tensor([0, 0, 1, -torch.mean(newd)])
+            else:
+                raise NotImplementedError
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:388]: NotImplementedError('xxxxx'), make the error mesage more clear

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 16]
```diff
@@ -0,0 +1,800 @@
...
+                                               pts_semantic_mask[indices][0]])
+
+            # Set the surface labels here
+            if self.primitive_mode == 'z':
+                if torch.sum(selected) > self.train_cfg['num_point'] and \
+                        torch.var(point2plane_dist[selected]) < \
+                        self.train_cfg['var_thresh']:
+                    center = points.new_tensor([
+                        (xmin + xmax) / 2.0, (ymin + ymax) / 2.0,
+                        torch.mean(coords[selected][:, 2])
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:434]: torch.mean(coords[selected][:, 2]) -> coords[selected][:, 2].mean()

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 17]
```diff
@@ -0,0 +1,800 @@
...
+                            line_center - coords[selected][line_select]
+                        point_sem[indices[selected][line_select]] = \
+                            points.new_tensor([line_center[0], line_center[1],
+                                               line_center[2],
+                                               pts_semantic_mask[indices][0]])
+
+            # Set the surface labels here
+            if self.primitive_mode == 'z':
+                if torch.sum(selected) > self.train_cfg['num_point'] and \
+                        torch.var(point2plane_dist[selected]) < \
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:430]: torch.var(point2plane_dist[selected]) -> point2plane_dist[selected].var()

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 18]
```diff
@@ -0,0 +1,800 @@
...
+                        point_offset[indices[selected][line_select]] = \
+                            line_center - coords[selected][line_select]
+                        point_sem[indices[selected][line_select]] = \
+                            points.new_tensor([line_center[0], line_center[1],
+                                               line_center[2],
+                                               pts_semantic_mask[indices][0]])
+
+            # Set the surface labels here
+            if self.primitive_mode == 'z':
+                if torch.sum(selected) > self.train_cfg['num_point'] and \
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:429]: torch.sum(selected) -> selected.sum()

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 19]
```diff
@@ -0,0 +1,800 @@
...
+            min_dist = point2plane_dist.min()
+            selected = torch.abs(point2plane_dist -
+                                 min_dist) < self.train_cfg['dist_thresh']
+
+            # Get upper four lines
+            if self.primitive_mode == 'line':
+                point2_lines_matching = self.match_point2line(
+                    coords[selected], xmin, xmax, ymin, ymax)
+                for idx, line_select in enumerate(point2_lines_matching):
+                    if torch.sum(line_select) > \
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:456]: torch.sum(line_select) -> line_select.sum()

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 20]
```diff
@@ -0,0 +1,800 @@
...
+                            self.train_cfg['num_point_line']:
+                        point_mask[indices[selected][line_select]] = 1.0
+                        line_center = torch.mean(
+                            coords[selected][line_select], axis=0)
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:460]: torch.mean(coords[selected][line_select], axis=0) -> coords[selected][line_select].mean(axis=0)

Check other places, use as simple implementation as possible

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 21]
```diff
@@ -0,0 +1,800 @@
...
+                    sel_global = indices[selected]
+                    point_mask[sel_global] = 1.0
+                    point_sem[sel_global] = points.new_tensor([
+                        center[0], center[1], center[2], zmax - zmin,
+                        (pts_semantic_mask[indices][0])
+                    ])
+                    point_offset[sel_global] = center - coords[selected]
+
+            # Get the boundary points here
+            vec1 = cur_corners[0] - cur_corners[4]
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:592]: Several lines of code are duplicated, can we write smaller functions to reduce repeated code?

For example
```python
            vec1 = cur_corners[0] - cur_corners[4]
            vec2 = cur_corners[4] - cur_corners[5]
            surface_norm = torch.cross(vec1, vec2)
            surface_dis = -torch.dot(surface_norm, cur_corners[5])
```

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 22]
```diff
@@ -0,0 +1,800 @@
...
+                plane_back = points.new_tensor([
+                    plane_front_temp[0], plane_front_temp[1],
+                    plane_front_temp[2], -torch.mean(refined_distance)
+                ])
+            else:
+                raise NotImplementedError(
+                    'Normal vector of the plane should be horizontal!')
+
+            # Get the boundary points here
+            point2plane_dist = torch.abs(
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:614]:
```python
point2plane_dist = torch.abs(
                torch.sum(coords * plane_front[:3], 1) + plane_front[-1])
            min_dist = point2plane_dist.min()
            selected = torch.abs(point2plane_dist -
                                 min_dist) < self.train_cfg['dist_thresh']

```
For getting boundary points is duplicated.

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 23]
```diff
@@ -0,0 +1,882 @@
...
+        elif sample_mod == 'random':
+            # Random sampling from the votes
+            batch_size, num_seed = seed_points.shape[:2]
+            sample_indices = torch.randint(
+                0,
+                num_seed, (batch_size, self.num_proposal),
+                dtype=torch.int32,
+                device=seed_points.device)
+        else:
+            raise NotImplementedError
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:162]: Error message

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 24]
```diff
@@ -0,0 +1,882 @@
...
+                label of each batch.
+            pts_instance_mask (None | list[torch.Tensor]): Point-wise instance
+                label of each batch.
+            bbox_preds (dict): Predictions from forward of primitive head.
+
+        Returns:
+            tuple[torch.Tensor]: Targets of primitive head.
+        """
+        valid_gt_masks = list()
+        gt_num = list()
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:283]: gt_num seems to be unnecessary

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 25]
```diff
@@ -0,0 +1,882 @@
...
+            aggregated_points (torch.Tensor): The aggregated points
+                of vote stage.
+
+        Returns:
+            Dict: Predictions of primitive head, including center,
+                semantic size and semantic scores.
+        """
+
+        ret_dict = {}
+        net_transposed = predictions.transpose(2, 1)
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:623]: pred_transposed

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 26]
```diff
@@ -0,0 +1,882 @@
...
+        if self.primitive_mode != 'line':
+            size_xyz_reshape = primitive_semantic.view(
+                batch_size * num_proposal, -1, self.num_dims).contiguous()
+            size_loss = self.semantic_reg_loss(
+                size_xyz_reshape,
+                gt_primitive_semantic,
+                dst_weight=gt_primitive_mask.view(batch_size * num_proposal,
+                                                  1))[1]
+        else:
+            size_loss = torch.tensor(0).float().to(center_loss.device)
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:741]: torch.tensor(0, device=center_loss.device), float() seems to be unnecessary

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 27]
```diff
@@ -0,0 +1,882 @@
...
+
+        Returns:
+            Tuple: Primitive center and the prediction indices.
+        """
+        ind_normal = F.softmax(pred_flag, dim=1)
+        pred_indices = (ind_normal[:, 1, :] >
+                        self.surface_thresh).detach().float()
+        selected = (ind_normal[:, 1, :] <=
+                    self.surface_thresh).detach().float()
+        offset = torch.ones_like(center) * self.upper_thresh
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:766]: Do we need to set the devices of offset.

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py [hunk 28]
```diff
@@ -0,0 +1,882 @@
...
+                line_center = coords[line_select].mean(dim=0)
+                line_center[center_axis] = corners[:, center_axis].mean()
+                point_offset[indices[line_select]] = \
+                    line_center - coords[line_select]
+                point_sem[indices[line_select]] = \
+                    point_sem.new_tensor([line_center[0], line_center[1],
+                                          line_center[2], cls_label])
+        return point_mask, point_offset, point_sem
+
+    def _assign_primitive_z_targets(self, point_mask, point_offset, point_sem,
```
Comment (ZwwWayne) [inline mmdet3d/models/roi_heads/mask_heads/primitive_head.py:808]: Try to merge `_assign_primitive_xy_targets` and `_assign_primitive_z_targets` by adding an option

File: mmdet3d/models/roi_heads/mask_heads/primitive_head.py (hunk 27 above)
Comment (encore-zhou) [reply mmdet3d/models/roi_heads/mask_heads/primitive_head.py:766]: `input.torch.zeros_like(input)` is equivalent to `torch.zeros(input.size(), dtype=input.dtype, layout=input.layout, device=input.device)`

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**